pkg/trace/obfuscate/sql.go

```diff
@@ -209,26 +210,33 @@ func (f *groupingFilter) Reset() {
 // some elements such as comments and aliases and obfuscation attempts to hide sensitive information
 // in strings and numbers by redacting them.
 func (o *Obfuscator) ObfuscateSQLString(in string) (*ObfuscatedQuery, error) {
```

Comment:
There are way too many functions and methods which address obfuscation. Let's simply change the ones we already have and not add more.

Thus, please remove both `*WithOptions` methods and simply use `o.opts.SQLObfuscationConfig` directly.

Reply:
We opted to not use the `SQLObfuscationConfig` directly from the **receiving** struct because of reasons listed in this discussion https://github.com/DataDog/datadog-agent/pull/8556#issuecomment-874960317. For a quick rundown, the CGo binding can only initialize the obfuscator config **once**, meaning once we make/add the `SQLObfuscationConfig` we cannot change the behavior for future calls. So if the first integration instance (i.e MySQL) initializes the obfuscator with `quantize` option `on`, but the second integration (i.e Postgres) does not need it, they'll inherit the option anyways. Because of this requirement, we decided to pass `options` directly to the method so we can change the behavior individually.

As to why we introduced a new `WithOptions` method, it simply reduced the number of signatures we needed to change/update and providing default options was easier. I don't have any hard feelings on this, so I've updated it to where we just change the existing methods as requested. https://github.com/DataDog/datadog-agent/pull/8556/commits/ad91185a828e69ce18d08c286e510fb4ee6ffe32

Reply:
I understand, but that makes a mess of the API, and I think it is not acceptable for a method to take an argument which specifies something that is already available on the receiver structure as a field. It's utterly confusing, and while it does fix your issue, it is just a bug awaiting to happen outside that context. As an alternative, are you not able to create a function instead of a method, which takes all parameters as arguments? Basically not use the `*Obfuscator` type, just call a function, something like `obfuscate.ObfuscateSQL(sql string, opts ....)`

Reply:
Updated: https://github.com/DataDog/datadog-agent/pull/8556/commits/2354a436ec591c7cbf376a72af4ef4ee95704c6d